File: playwright/pages/InventoryPage.ts
import { Page, Locator } from '@playwright/test';

export class InventoryPage {
  readonly page: Page;
  readonly inventoryItems: Locator;
  readonly shoppingCartBadge: Locator;
  readonly sortDropdown: Locator;
  readonly cartLink: Locator;

  constructor(page: Page) {
    this.page = page;
    this.inventoryItems = page.locator('.inventory_item');
    this.shoppingCartBadge = page.locator('.shopping_cart_badge');
    this.sortDropdown = page.locator('[data-test="product-sort-container"]');
    this.cartLink = page.locator('.shopping_cart_link');
  }

  async addItemToCart(itemName: string) {
    const item = this.page.locator('.inventory_item', { hasText: itemName });
    await item.locator('button.btn_inventory').click();
  }

  async removeItemFromCart(itemName: string) {
    const item = this.page.locator('.inventory_item', { hasText: itemName });
    await item.locator('button.btn_inventory').click();
  }

  async getCartCount() {
    const badge = this.shoppingCartBadge;
    if (await badge.isVisible()) {
      return parseInt(await badge.textContent() || '0', 10);
    }
    return 0;
  }

  async sortBy(option: string) {
    await this.sortDropdown.selectOption({ label: option });
  }

  async getItemNames(): Promise<string[]> {
    return this.page.locator('.inventory_item_name').allTextContents();
  }

  async getItemPrices(): Promise<number[]> {
    const prices = await this.page.locator('.inventory_item_price').allTextContents();
    return prices.map(p => parseFloat(p.replace('$', '')));
  }

  async goToCart() {
    await this.cartLink.click();
  }
}

File: bdd-cucumber/src/test/java/pages/InventoryPage.java
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.Select;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class InventoryPage {
    private final WebDriver driver;

    private static final By TITLE = By.cssSelector(".title");
    private static final By INVENTORY_ITEMS = By.cssSelector(".inventory_item");
    private static final By ITEM_NAMES = By.cssSelector(".inventory_item_name");
    private static final By ITEM_PRICES = By.cssSelector(".inventory_item_price");
    private static final By ITEM_IMAGES = By.cssSelector(".inventory_item_img");
    private static final By CART_BADGE = By.cssSelector(".shopping_cart_badge");
    private static final By CART_LINK = By.cssSelector(".shopping_cart_link");
    private static final By SORT_SELECT = By.cssSelector(".product_sort_container");

    public InventoryPage(WebDriver driver) {
        this.driver = driver;
    }

    public String getTitle() {
        return driver.findElement(TITLE).getText();
    }

    public int getItemCount() {
        return driver.findElements(INVENTORY_ITEMS).size();
    }

    public void addItemToCart(String itemName) {
        String xpath = String.format("//div[text()='%s']/ancestor::div[@class='inventory_item']//button[starts-with(@id,'add-to-cart')]", itemName);
        driver.findElement(By.xpath(xpath)).click();
    }

    public void removeItemFromCart(String itemName) {
        String xpath = String.format("//div[text()='%s']/ancestor::div[@class='inventory_item']//button[starts-with(@id,'remove')]", itemName);
        driver.findElement(By.xpath(xpath)).click();
    }

    public int getCartBadgeCount() {
        var badges = driver.findElements(CART_BADGE);
        return badges.isEmpty() ? 0 : Integer.parseInt(badges.get(0).getText());
    }

    public boolean isCartBadgeVisible() {
        return driver.findElements(CART_BADGE).size() > 0;
    }

    public void goToCart() {
        driver.findElement(CART_LINK).click();
    }

    public void sortBy(String sortOption) {
        var select = new Select(driver.findElement(SORT_SELECT));
        select.selectByVisibleText(sortOption);
    }

    public List<String> getItemNames() {
        return driver.findElements(ITEM_NAMES)
                .stream()
                .map(WebElement::getText)
                .collect(Collectors.toList());
    }

    public List<Double> getItemPrices() {
        return driver.findElements(ITEM_PRICES)
                .stream()
                .map(e -> e.getText().replace("$", ""))
                .map(Double::parseDouble)
                .collect(Collectors.toList());
    }

    public List<String> getBrokenImageSrc() {
        var images = driver.findElements(ITEM_IMAGES);
        List<String> broken = new ArrayList<>();
        for (var img : images) {
            var src = img.getAttribute("src");
            if (src == null || src.contains("invalid") || src.contains("broken") || src.isEmpty()) {
                broken.add(src);
            }
        }
        return broken;
    }
}
